C7Engine/EntryPoints/MessageToEngine.cs

```diff
@@ -1,149 +1,164 @@
 
-public abstract class MessageToEngine {
-	public abstract void process();
+		public void send()
+		{
+			EngineStorage.pendingMessages.Enqueue(this);
+			EngineStorage.actionAddedToQueue.Set();
+		}
+	}
 
-	public void send()
-	{
-		EngineStorage.pendingMessages.Enqueue(this);
-		EngineStorage.actionAddedToQueue.Set();
+	public class MsgShutdownEngine : MessageToEngine {
+		public override void process()
+		{
+			Console.WriteLine("Engine received shutdown message.");
+		}
 	}
-}
 
-public class MsgShutdownEngine : MessageToEngine {
-	public override void process()
+	public class MsgSetFortification : MessageToEngine
 	{
-		Console.WriteLine("Engine received shutdown message.");
-	}
-}
+		private string unitGUID;
+		private bool fortifyElseWake;
 
-public class MsgSetFortification : MessageToEngine
-{
-	private string unitGUID;
-	private bool fortifyElseWake;
+		public MsgSetFortification(string unitGUID, bool fortifyElseWake)
+		{
+			this.unitGUID = unitGUID;
+			this.fortifyElseWake = fortifyElseWake;
+		}
 
-	public MsgSetFortification(string unitGUID, bool fortifyElseWake)
-	{
-		this.unitGUID = unitGUID;
-		this.fortifyElseWake = fortifyElseWake;
+		public override void process()
+		{
+			MapUnit unit = EngineStorage.gameData.GetUnit(unitGUID);
+
+			// Simply do nothing if we weren't given a valid GUID. TODO: Maybe this is an error we need to handle? In an MP game, we should reject
+			// invalid actions at the server level but at the client level an invalid action received from the server indicates a desync.
+			if (unit != null) {
+				if (fortifyElseWake)
+					unit.fortify();
+				else
+					unit.wake();
+			}
+		}
 	}
 
-	public override void process()
+	public class MsgMoveUnit : MessageToEngine
 	{
-		MapUnit unit = EngineStorage.gameData.mapUnits.Find(u => u.guid == unitGUID);
+		private string unitGUID;
+		private TileDirection dir;
 
-		// Simply do nothing if we weren't given a valid GUID. TODO: Maybe this is an error we need to handle? In an MP game, we should reject
-		// invalid actions at the server level but at the client level an invalid action received from the server indicates a desync.
-		if (unit != null) {
-			if (fortifyElseWake)
-				unit.fortify();
-			else
-				unit.wake();
+		public MsgMoveUnit(string unitGUID, TileDirection dir)
+		{
+			this.unitGUID = unitGUID;
+			this.dir = dir;
 		}
-	}
-}
 
-public class MsgMoveUnit : MessageToEngine
-{
-	private string unitGUID;
-	private TileDirection dir;
-
-	public MsgMoveUnit(string unitGUID, TileDirection dir)
-	{
-		this.unitGUID = unitGUID;
-		this.dir = dir;
+		public override void process()
+		{
+			MapUnit unit = EngineStorage.gameData.GetUnit(unitGUID);
+			unit?.move(dir);
+		}
 	}
 
-	public override void process()
+	public class MsgSetUnitPath : MessageToEngine
 	{
-		MapUnit unit = EngineStorage.gameData.mapUnits.Find(u => u.guid == unitGUID);
-		if (unit != null)
-			unit.move(dir);
-	}
-}
+		private string unitGUID;
+		private Tile dest;
```

Comment:
Makes sense thank you for pointing this out!

I haven't read forum discussions on GUIDs, but I am definitely in favour of not using GUIDs. I'm a fan of auto-incrementing integer IDs. They could also be scoped by an accompanying class ID to simplify generating unique IDs for each object. Do we have a thread specifically discussing this? Whatever we use it's probably something worth figuring out early on

Reply:
I agree that auto-incrementing integer IDs, i.e. serial numbers, are the way to go. There's no thread about this on the forums, it was discussed briefly in the architecture thread starting with my post here: https://forums.civfanatics.com/threads/dev-architecture-thread.674254/#post-16218946. We ought to start such a thread to get this settled once and for all. The way I see it, the hard part about identifying objects is how those IDs will work with mods. In particular, mods should be able to define their own types of game objects, which will be given serial numbers and a class ID, and players should be able to add or remove mods from a saved game and have it still work, within reason. It will require some care to make sure we can still reliably identify objects as the ground moves beneath us.

Reply:
I'll start a [discussion](https://github.com/C7-Game/Prototype/discussions) about it so we can also try out GitHub discussions. I'd like to merge this PR without adding Tile GUIDs, and refactor the GoTo functionality to use whatever we settle on, since AFAIK tiles don't have GUIDs currently.

Reply:
Tiles don't have GUIDs, that's why I suggested using the tile index or coordinates. But it's okay if you want to merge this PR without removing the object references since they will only be a problem for multiplayer, which is still a ways off.